nunaliit2-couch-client/src/main/java/ca/carleton/gcrc/couch/client/impl/listener/ModuleMetadataChangeListener.java

```diff
@@ -0,0 +1,165 @@
+package ca.carleton.gcrc.couch.client.impl.listener;
+
+import ca.carleton.gcrc.couch.client.CouchDb;
+import ca.carleton.gcrc.couch.client.CouchDesignDocument;
+import ca.carleton.gcrc.couch.client.CouchQuery;
+import ca.carleton.gcrc.couch.client.CouchQueryResults;
+import ca.carleton.gcrc.exception.NunaliitException;
+import org.apache.commons.lang3.tuple.Pair;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Stores module metadata in memory, updates this data if the module is changed in the database.
+ */
+public final class ModuleMetadataChangeListener extends AbstractCouchDbChangeListener
+{
+    private static final Logger logger = LoggerFactory.getLogger(ModuleMetadataChangeListener.class);
+    private static final Object lockObj = new Object();
+    private final CouchDb couchDb;
+    private final Set<String> moduleIds;
+    private final Map<String, JSONObject> moduleMetadata;
+
+    public ModuleMetadataChangeListener(CouchDb couchDb) throws NunaliitException {
+        super(couchDb);
+
+        this.couchDb = couchDb;
+        moduleIds = Collections.synchronizedSet(new HashSet<String>());
+        moduleMetadata = Collections.synchronizedMap(new HashMap<String, JSONObject>());
+    }
+
+    /**
+     * Request metadata for the given module Id. If the metadata is not stored in memory, will add the module to watch
+     * list and update metadata if it exists. This is how we start tracking new module metadata, only if their
+     * metadata is requested.
+     *
+     * @param moduleId The module to fetch metadata for.
+     * @return The contents of "nunaliit_metdata" in the module document.
+     */
+    public JSONObject findMetadata(String moduleId) {
+        JSONObject metadata;
+
+        // Need to add it
```

Comment:
It's better to synchronized the whole function on lockObj. Because the two steps: 1. Check if moduleid exists and 2. Add new moduleId if it is not present, need to be an atomic operation as a whole together.

Reply:
Good catch. Fixed.